src/gcal.ts

```diff
@@ -37,83 +43,104 @@ const calendar = google.calendar({
-  const p = () =>
-    new Promise<Credentials>((resolve, reject) => {
-      const retrieveCredentials = () => {
-        if (fs.existsSync(credentialsPath)) {
-          const tokens = JSON.parse(
-            fs.readFileSync(credentialsPath, "utf-8")
-          ) as Credentials;
-          if (tokens.expiry_date > Date.now()) {
-            oauth2Client.setCredentials(tokens);
-            return tokens;
-          }
-        }
-        return null;
-      };
-
-      const credentials = retrieveCredentials();
-      if (credentials) {
-        resolve(credentials);
-        return;
+function retrieveExistingCredentials(): TaskEither<
+  unknown,
+  Option<Credentials>
+> {
+  return taskEither.fromIO(() => {
+    if (fs.existsSync(credentialsPath)) {
+      const tokens = JSON.parse(
+        fs.readFileSync(credentialsPath, "utf-8")
+      ) as Credentials;
+      if (tokens.expiry_date > Date.now()) {
+        oauth2Client.setCredentials(tokens);
+        return fromNullable(tokens);
       }
+    }
+    return none;
+  });
+}
 
-      const scopes = [
-        "https://www.googleapis.com/auth/calendar.events",
-        "https://www.googleapis.com/auth/userinfo.profile"
-      ];
-
-      const url = oauth2Client.generateAuthUrl({
-        access_type: "offline",
-        scope: scopes
-      });
-
-      logInfo("\nVisit this URL to authenticate with your Google account:");
-      logDetail(`\n  ${url}`);
-
-      const app = express();
-
-      app.get("/authorize", (req, res) => {
-        const code = req.query.code;
-        res
-          .status(200)
-          .send(
-            "Successfully authenticated! You can now close this tab and go back to your terminal."
-          );
-
-        oauth2Client.getToken(code, (err, tokens) => {
-          if (!err) {
-            if (!fs.existsSync(ferioDir)) {
-              fs.mkdirSync(ferioDir, { recursive: true });
-            }
-            fs.writeFileSync(credentialsPath, JSON.stringify(tokens, null, 2));
-            oauth2Client.setCredentials(tokens);
-            logInfo(
-              "\n:white_check_mark: Successfully authenticated. Let's move on!"
-            );
-            resolve(tokens);
-          } else {
-            logError(`${err.response.status} ${err.response.statusText}`);
-            logError(JSON.stringify(err.response.data));
-            reject(err);
-          }
-        });
-      });
-
-      app.listen(port);
-    });
-
-  return tryCatch(p, r => r);
+function exchangeCodeForCredentials(
+  code: GetTokenOptions
+): TaskEither<unknown, Credentials> {
+  return fluent(taskEither)(taskify(oauth2Client.getToken)(code))
+    .chain(
+      credentials => saveCredentials(credentials, oauth2Client) as any // TODO
+    )
+    .mapLeft(err =>
+      fluent(taskEither)(
+        logError(`${err.response.status} ${err.response.statusText}`)
+      )
+        .chain(() => logError(JSON.stringify(err.response.data)))
+        .value()
```

Comment:
@gcanti sarebbe comodo avere qui un `chainLeft` di qualche tipo. `logError` torna un `TaskEither` e qui ho dovuto runnarlo a mano.

Reply:
@gabro non capisco bene l'operazione, se `logError` torna un `TaskEither` quella `mapLeft` mi pare strana

Reply:
sostanzialmente voglio loggare in caso di errore, ma l'operazione di log torna un TaskEither

Reply:
ok ma la firma come sarebbe?

```ts
declare function chainLeft<A, L, M, B>(ma: TaskEither<L, A>, f: (l: L) => TaskEither<M, B>): TaskEither<?, ?>
```